### DISCUSSION TITLE
Possibbility to clear whole chat to remove all messages / destroy chat room

### DISCUSSION THREAD

**Discussion by rugk** [element-hq/element-meta#850] [General]
**Is your suggestion related to a problem? Please describe.**
I'm always frustrated when I want to clear all messages in a chat to just remove the whole history, and to delete my data.
It may be encrypted etc., but deleted is better than encrypted.

Just so that I may want to "forget" (but technically secure, i.e. delete) things I have written once. Just as oral communication is not always permanent.

I know I could choose each of my message and delete it one-by-one, but that is obviously not possible in large chats.

**Describe the solution you'd like**
A button in the menu:
* "Clear whole chat."
* When choosing it a big warning that makes me aware that it really **deletes** all the data. Maybe even make me enter some random word (or the name of the chat/room) you choose, or so like when you try to delete a GitHub repo.
* Delete whole chat history **for me only**. (Because of course I cannot force other people to delete their history.) 

**Describe alternatives you've considered**

I know I can leave the room and all other persons too, but AFAIK the room is still not deleted (automatically) then. (Sorry, cannot find the corresponding issue right now.)

Implementation alternative: If that is too much for rooms with many members, I guess private chats which are affected by this may only likely be 1:1 (DM) and e2e-encrypted chats.
So you could also limit deleting messages to these chats.
**Possibly** you could also make all members in a room/chat top agree to a deletion and then delete it for _all of them_, i.e. "consensus"-like. (Of course _everyone_ has to agree.) Button could be named "Request chat deletion [from members]".
(In my use cases likely all members want it to be deleted. If only one wants it, they can just leave the room, because then they likely just "don't want to see it". This is a different use case than "I/We want it to be deleted." though, see GDPR below.)

Also if you want to highlight the destructiveness of this, you may also call it "destroy room".

**Additional context**
This may be GDPR-relevant, because it includes the "right to be forgotten", i.e. you may delete your data at any time.
Possibly, you can do so for your whole account already, but IMHO, it also implied you should be able to delete a single chat.

**Comment by t3chguy** [comment]
> This may be GDPR-relevant, because it includes the "right to be forgotten", i.e. you may delete your data at any time.

Given that Matrix fits within the same category as e-mail this isn't quite true. When you send a message you are authorizing every other party a copy of that message and they are actually able to keep it for as long as they wish. Just like if you GDPR Gmail that won't remove any emails which you sent to your friends over at *@hotmail.com

**Reply by evo11x** [reply]
> > This may be GDPR-relevant, because it includes the "right to be forgotten", i.e. you may delete your data at any time.
> 
> Given that Matrix fits within the same category as e-mail this isn't quite true. When you send a message you are authorizing every other party a copy of that message and they are actually able to keep it for as long as they wish. Just like if you GDPR Gmail that won't remove any emails which you sent to your friends over at *@hotmail.com

I can always delete my emails from inbox, so there should be a delete option in element to delete my messages.

**Comment by rugk** [comment]
Fine, but it's more of a ethically aspect, as if **all parties** want the messages removed, that is currently very hard. So it's not that only one party wants it to be removed. That is nto the use case, I intended to describe.

**Comment by t3chguy** [comment]
This really belongs on matrix-doc, it is related to https://github.com/matrix-org/matrix-doc/issues/1882

**Comment by rugk** [comment]
Okay, created https://github.com/matrix-org/matrix-doc/issues/2686.

Though depending on the solution, this may also be a client-only thing, i.e. *delete messages only for me in this chat*, so may not be a matrix thing.
As such leaving this open. If you disagree, feel free to explain why and close this issue.

**Comment by dtluna** [comment]
Me and my friends want to clean our private chat for security reasons regularly.

**Comment by t3chguy** [comment]
You could set up a message retention policy https://github.com/matrix-org/synapse/blob/babolivier/retention_doc/docs/message_retention_policies.md

**Comment by ell1e** [comment]
> You could set up a message retention policy

How would one do this in element? This sounds useful. (Also, the link appears to be dead, but I assume it's some "delete older than date X" mechanism.)

**Comment by dklimpel** [comment]
There you can find the information: https://matrix-org.github.io/synapse/latest/message_retention_policies.html

**Reply by ell1e** [reply]
This only shows server-side info, but what's the UI in element to do this for regular users? Not everybody knows or would like to know how to admin a server to access a basic delete old log feature. My apologies if I'm missing something obvious. Edit: also, this functionality should be accessible without having any admin permissions as well, if it concerns only your own messages.

**Comment by xyonium** [comment]
> There you can find the information: https://matrix-org.github.io/synapse/latest/message_retention_policies.html

I tried retention policy, but it only clear the data from the server side. The client need to manually delete everything. Is this the intended behavior or am I doing something wrong?

**Comment by ThePharaohArt** [comment]
Surprised this has not been implemented yet. There current exists "delete recent messages" under admin tools under a specified member, but still no "delete all messages" in a DM.

**Comment by XJY00** [comment]
I have the same need

**Comment by d668** [comment]
In case someone needs, this script if run under room admin, deletes all messages. I wonder is there a better solution? 

```
#!/bin/bash

# Room ID and updated access token
ROOM_ID="!xx:matrix.xx.xx"
ACCESS_TOKEN="xxx"
SERVER_URL="https://matrix.xx.xx"

# Fetching all event IDs (simplified example, pagination may be required for large rooms)
response=$(curl -s -X GET \
  -H "Authorization: Bearer $ACCESS_TOKEN" \
  "$SERVER_URL/_matrix/client/r0/rooms/$ROOM_ID/messages?dir=b&limit=1000")

echo "Response: $response"  # Print the response for debugging

event_ids=$(echo "$response" | jq -r '.chunk[].event_id')

# Check if event_ids is not null
if [ -z "$event_ids" ]; then
  echo "No events found or there was an issue with the request."
  exit 1
fi

# Loop through each event and redact it
for event_id in $event_ids
do
  echo "Redacting event $event_id"
  curl -X PUT \
    -H "Authorization: Bearer $ACCESS_TOKEN" \
    -H "Content-Type: application/json" \
    -d '{"reason": "Admin deletion"}' \
    "$SERVER_URL/_matrix/client/r0/rooms/$ROOM_ID/redact/$event_id/$(uuidgen)"
done
```

**Reply by rugk** [reply]
1. You have posted this twice.
2. FYI, [you can also mark code snippets with the language, which adds syntax highlighting to them, in the Markdown you write on GitHub](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks).

@d668

**Comment by d668** [comment]
this is the script that deletes all messages from all rooms. User should be admin and in the room

```
#!/bin/bash

# Updated access token
ACCESS_TOKEN="xxx"
SERVER_URL="https://matrix.xxx.xxx"

# Set the threshold duration in milliseconds (e.g., 86400000 for 1 day)
DURATION_MS=86400000 # Replace with the desired duration in milliseconds

# Calculate the threshold timestamp: current time minus the duration
THRESHOLD_TIMESTAMP=$(($(date +%s%3N) - DURATION_MS))

# Fetching the list of all rooms the user has access to
rooms=$(curl -s -X GET \
  -H "Authorization: Bearer $ACCESS_TOKEN" \
  "$SERVER_URL/_matrix/client/r0/joined_rooms" | jq -r '.joined_rooms[]')

# Loop through each room and process events
for ROOM_ID in $rooms; do
  echo "Processing room: $ROOM_ID"
  
  # Fetching all event IDs and their timestamps for the current room
  response=$(curl -s -X GET \
    -H "Authorization: Bearer $ACCESS_TOKEN" \
    "$SERVER_URL/_matrix/client/r0/rooms/$ROOM_ID/messages?dir=b&limit=1000")

  echo "Response: $response"  # Print the response for debugging

  # Parse event IDs and timestamps
  event_data=$(echo "$response" | jq -c '.chunk[] | {event_id: .event_id, timestamp: .origin_server_ts}')

  # Check if event_data is not null
  if [ -z "$event_data" ]; then
    echo "No events found or there was an issue with the request."
    continue
  fi

  # Loop through each event and redact it if it's older than the threshold
  echo "$event_data" | while IFS= read -r event; do
    event_id=$(echo "$event" | jq -r '.event_id')
    timestamp=$(echo "$event" | jq -r '.timestamp')

    if [ "$timestamp" -lt "$THRESHOLD_TIMESTAMP" ]; then
      echo "Redacting event $event_id with timestamp $timestamp in room $ROOM_ID"
      curl -X PUT \
        -H "Authorization: Bearer $ACCESS_TOKEN" \
        -H "Content-Type: application/json" \
        -d '{"reason": "Admin deletion"}' \
        "$SERVER_URL/_matrix/client/r0/rooms/$ROOM_ID/redact/$event_id/$(uuidgen)"
    else
      echo "Skipping event $event_id with timestamp $timestamp in room $ROOM_ID"
    fi
  done
done

```

**Reply by d668** [reply]
it does delete the messages but the room becomes unusable after that. Sucks

**Reply by ell1e** [reply]
So what do users do that aren't admins, for their own messages?

This is such a basic feature, that it's missing from a messenger advertised for security in particular seems like an unfortunate mismatch. Basic security practices including not storing leakable data for longer than needed, which requires easy mass deletion controls.

**Reply by sushantdhiman** [reply]
open sores /s

**Comment by hooonza** [comment]
What about this?   https://www.sami-lehtinen.net/blog/matrix-auto-redact-bot
